pyracf/common/logger.py

```diff
@@ -93,33 +95,65 @@ def log_debug(self, header_message: str, message: str) -> None:
-    ) -> str:
-        """Redact a list of strings or sequences of bytes in a string or bytes object"""
-        if not redact_strings:
-            return string
-        for to_redact in redact_strings:
-            if isinstance(string, bytes):
-                string = self.redact_string(
-                    string, redact_string=bytes(to_redact, "utf-8")
-                )
-            else:
-                string = self.redact_string(string, redact_string=to_redact)
-        return string
+        dictionary: dict,
+        secret_traits: dict,
+    ) -> dict:
+        """Redact a list of specified secret traits in a request dictionary"""
+        for seg_trait in secret_traits:
+            segment = seg_trait.split(":")[0]
+            if not dictionary.get(segment, {}).get(seg_trait, {}).get("value", None):
+                continue
+            dictionary[segment][seg_trait]["value"] = "********"
+        return dictionary
+
+    def __redact_string(
+        self,
+        input_string: Union[str, bytes],
+        start_ind: int,
+        end_pattern: str,
+    ):
+        """Redacts characters in a string between a starting index and ending pattern"""
+        """Replaces the identified characters with ******** (regardless of length)"""
+        pre_keyword = input_string[:start_ind]
+        post_keyword = end_pattern.join(input_string[start_ind:].split(end_pattern)[1:])
+        return pre_keyword + "********" + end_pattern + post_keyword
 
-    def redact_string(
-        self, string: Union[str, bytes], redact_string: Union[str, bytes]
+    def redact_request_xml(
+        self,
+        xml_string: Union[str, bytes],
+        secret_traits: dict,
+    ) -> Union[str, bytes]:
+        """Redact a list of specific secret traits in a request xml string or bytes object"""
+        """Bases this on xml pattern <xml_key any-operation>xml_value</xml_key>"""
+        is_bytes = False
+        if isinstance(xml_string, bytes):
+            is_bytes = True
+            xml_string = xml_string.decode("utf-8")
+        for xml_key in secret_traits.values():
+            match = re.search(rf"\<{xml_key}+[^>]*\>", xml_string)
+            if not match:
+                continue
+            xml_string = self.__redact_string(xml_string, match.end(), f"</{xml_key}")
+        if is_bytes:
+            xml_string = xml_string.encode("utf-8")
+        return xml_string
+
+    def redact_result_xml(
+        self,
+        xml_string: str,
+        secret_traits: dict,
     ) -> str:
-        """Redact a string or sequence of byte in a bytes object."""
-        if not redact_string:
-            return string
-        redacted_string = "********"
-        if isinstance(redact_string, bytes):
-            redacted_string = b"********"
-        return string.replace(redact_string, redacted_string)
+        """Redacts a list of specific secret traits in a result xml string"""
+        """Bases this on RACF command pattern RACF_KEY( value ) """
```

Comment:
```python
        """
        Redacts a list of specific secret traits in a result xml string.
        Based on the following RACF command pattern:
            'TRAIT (value)'
        """
```

Also, I am guessing that this ignores white space since there is always a lot of white space in the RACF commands returned by IRRSMO00. In this case, I think it would be helpful to add a comment about this in the docstring.

Reply:
Now changed to:

```python
        """
        Redacts a list of specific secret traits in a result xml string.
        Based on the following RACF command pattern:
            'TRAIT (value)'
        This function also accounts for varied amounts of whitespace in the pattern.
        """
```